### PR TITLE
Feature/27 batch-processing support

### PR DESCRIPTION
**Pull request by zsinnema**
add batch-processing support
see #27

### CODE REVIEW

**Comment by gidili** [comment]
Hey @phildong this is ready for you to review

File: pipeline.ipynb [hunk 1]
```diff
@@ -973,10 +885,10 @@
     "hv.output(size=output_size)\n",
     "opts_im = dict(plot=dict(height=b_init.sizes['height'], width=b_init.sizes['width'], colorbar=True), style=dict(cmap='Viridis'))\n",
     "opts_cr = dict(plot=dict(height=b_init.sizes['height'], width=b_init.sizes['height'] * 2))\n",
-    "(regrid(hv.Image(b_init, kdims=['width', 'height'])).opts(**opts_im).relabel('Background Spatial Initial')\n",
-    " + datashade(hv.Curve(f_init, kdims=['frame'])).opts(**opts_cr).relabel('Background Temporal Initial')\n",
-    " + regrid(hv.Image(b_spatial, kdims=['width', 'height'])).opts(**opts_im).relabel('Background Spatial First Update')\n",
-    " + datashade(hv.Curve(f_spatial, kdims=['frame'])).opts(**opts_cr).relabel('Background Temporal First Update')\n",
+    "(regrid(hv.Image(b_init.compute(), kdims=['width', 'height'])).opts(**opts_im).relabel('Background Spatial Initial')\n",
```
Comment (phildong) [inline pipeline.ipynb:888]: maybe separate this out since we have another PR for this?

**Review by phildong — COMMENTED** [review]
I'm not sure I understand why `dpath` is special and need to be set with a separate `globals.py`, can we get around that?
Also I'm hesitating how friendly this is to non-batch-processing cases, since now if I want to see my current parameter and potentially change them, I have to call them in a terminal and modify a separate notebook. Maybe @wmau can chime in on this?

**Comment by wmau** [comment]
I am not familiar with `import_ipynb`. Are you able to modify the `config.ipynb` parameters in `pipeline.ipynb` after you do `import config`? If so, it would work the same way as it currently does more or less @phildong. 

But I agree that I don't see any reason why `dpath` needs to be a global variable since it's always passed as an argument to every function that I know of.

**Comment by zsinnema** [comment]
> I'm not sure I understand why `dpath` is special and need to be set with a separate `globals.py`, can we get around that?

** dpath is "strange" because it's reused in the save param list so we need to pass it to config.ipynb


> Also I'm hesitating how friendly this is to non-batch-processing cases, since now if I want to see my current parameter and potentially change them, I have to call them in a terminal and modify a separate notebook. Maybe @wmau can chime in on this?

** you can easily open the config.ipynb notebook to see all params and explanation about the params

I do believe that having a "global" config file for multiple notebooks is a good approach

**Comment by zsinnema** [comment]
> I am not familiar with `import_ipynb`. Are you able to modify the `config.ipynb` parameters in `pipeline.ipynb` after you do `import config`? If so, it would work the same way as it currently does more or less @phildong.

** with this setup you can reuse the config in different notebooks which could be very handy

> 
> But I agree that I don't see any reason why `dpath` needs to be a global variable since it's always passed as an argument to every function that I know of.

** see my previous comment

**Comment by wmau** [comment]
So you cannot modify any parameters after importing `config.ipynb` without opening that notebook again? This would make the workflow very cumbersome.

**Comment by zsinnema** [comment]
> So you cannot modify any parameters after importing `config.ipynb` without opening that notebook again? This would make the workflow very cumbersome.

** no not exactly. You can just overwrite "as usual" there is no need to re-open the config notebook.
after importing the config notebook all varaiables are present in the current notebook like they are currently. there is no difference

**Comment by wmau** [comment]
In that case, I think this could work @phildong , because the user could still import `config.ipynb` and change parameters inside `pipeline.ipynb`. I still don't think `dpath` needs to be global though. `param_save_minian` should get its own path because users may not necessarily want to save their minian outputs in the same folder as their data.

**Comment by phildong** [comment]
my only concern is that before we have a "reference" of the default values and other parameters out there within the notebook, and at least for me every time I want to change something I change them at those "default" places, instead of inserting a cell at random locations, and I feel that keeps me organized. If this is not a typical workflow or you think there are some easy ways to achieve something similar after this separation then I think it can totally work.

**Comment by gidili** [comment]
Maybe we could have a cell printing the params out from config and
suggesting to the user to override any of them at the top, so we have both
a display of values and centralised default configuration

**Comment by zsinnema** [comment]
there are many parameter, are they all being "tweaked" by the end user or are there only some?
we could add a "comment" in the step show the most likely params to tweak (and add comments about what they do and how to tweak them)

Or may be better add them at/in the correct context, so the most important (tweakable) params for step x are explained and overwritable at step x

@phildong what do you think

**Comment by phildong** [comment]
So we've just had a discussion on this, I believe the main issue with this approach is that there lack an elegant way to record user-made changes that's beyond changing parameters, and that's mainly due to the fact that minian is constantly changing and user's data are highly heterogeneous, and there is a non-negligible chance that users would like to modify the pipeline/code itself. The only consistent way we can think of to record these information is to save a complete copy of the pipeline notebook including all the actually-used parameters for each session, even for batch processing, and if we are going in this route it seems separating parameters from the pipeline only make things more redundant than streamlined.

So for now I think the main focus on batch-processing support would be to restore the functionality to previous states, namely programatically executing a notebook for multiple sessions. If we're using `papermill` for this then the main challenge is to make the parameters friendly to json (see #27)

However if we can find a more elegant way to record both the parameters and the pipeline/custom code for the analyzed datasets than what we currently have then we're definitely open to it!

@wmau feel free to add anything that I missed!
